src/Processors/FractionalLimitTransform.cpp

```diff
@@ -172,6 +185,66 @@ FractionalLimitTransform::Status FractionalLimitTransform::pullData(PortsData &
 
     chunks_cache.push_back({data.output_port, std::move(data.current_chunk)});
 
+    /// This optimizes if offset_fraction is set but the above block optimizes if integral offset is set and both can't be non-zero.
+    ///
+    /// Detect blocks that will 100% get removed by the fractional offset and remove them as early as possible.
+    /// example: if we have 10 blocks with same num of rows and offset 0.1 we can freely drop the first block even before reading all data.
+    while (!chunks_cache.empty() && std::ceil(rows_cnt * offset_fraction) - rows_read_from_cache >= chunks_cache.front().chunk.getNumRows())
+    {
+        rows_read_from_cache += chunks_cache.front().chunk.getNumRows();
+        chunks_cache.pop_front();
+    }
+
+    /// If offset_fraction is zero we can detect blocks that
+    /// will 100% get pushed and push them as early as possible.
+    if (offset_fraction == 0 && !chunks_cache.empty())
+    {
+        auto output = *chunks_cache.front().output_port;
+        /// Check can output?
+        if (output.isFinished())
+        {
+            input.close();
+            return Status::Finished;
+        }
+        if (!output.canPush())
+            return Status::PortFull;
+
+        auto & cache_chunk = chunks_cache.front().chunk;
+        auto num_rows = cache_chunk.getNumRows();
+
+        UInt64 remaining_offset = 0;
+        if (rows_read_from_cache < offset)
+            remaining_offset = offset - rows_read_from_cache;
+
+        auto curr_limit = std::ceil(rows_cnt * limit_fraction);
```

Comment:
```suggestion
        UInt64 curr_limit = std::ceil(rows_cnt * limit_fraction);
```
Otherwise, `cur_limit` would have type `double` which could potentially create problems later.